File: python/conway.py
from os import system
from random import random
from time import sleep

DELTAS = [[0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1], [-1, 0], [-1, 1]]


class Conway:
    def __init__(self, width: int, height: int) -> None:
        self.width = width
        self.height = height
        self.grid = self.create_new_grid()

    def create_new_grid(self):
        grid = []
        for _ in range(self.width):
            column = []
            for _ in range(self.height):
                cell = "@" if random() < 0.5 else " "
                column.append(cell)
            grid.append(column)
        return grid

    def print_grid(self):
        for y in range(self.height):
            for x in range(self.width):
                print(self.grid[x][y], end="")
            print()

    def next_generation(self):
        next_grid = []
        for x in range(self.width):
            next_column = []
            for y in range(self.height):
                n = self.count_neighbors(x, y)
                if self.grid[x][y] == "@":
                    next_column.append("@" if 2 <= n <= 3 else " ")
                else:
                    next_column.append("@" if n == 3 else " ")
            next_grid.append(next_column)
        self.grid = next_grid

    def count_neighbors(self, x, y):
        neighbors = 0
        for d in DELTAS:
            nx = x + d[0]
            nx = self.width - 1 if nx < 0 else nx
            nx = 0 if nx > self.width - 1 else nx

            ny = y + d[1]
            ny = self.height - 1 if ny < 0 else ny
            ny = 0 if ny > self.height - 1 else ny

            if self.grid[nx][ny] == "@":
                neighbors += 1
        return neighbors

    def play_game(self):
        try:
            while True:
                system("clear")
                self.print_grid()
                sleep(0.5)
                self.next_generation()
        except KeyboardInterrupt:
            SystemExit()


if __name__ == "__main__":
    conway = Conway(80, 24)
    conway.play_game()

File: c/conway.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define ROWS 24
#define COLS 80

const int deltas[8][2] = {
	{0, 1}, {1, 1}, {1, 0}, {1, -1},
	{0, -1}, {-1, -1}, {-1, 0}, {-1, 1}
};

int count_neighbors(int x, int y, char grid[ROWS][COLS]);
void next_generation(char grid[ROWS][COLS]);
void print_grid(char grid[ROWS][COLS]);

int main(void)
{
	int x, y;
	char grid[ROWS][COLS];

	for (x = 0; x < ROWS; x++) {
		for (y = 0; y < COLS; y++) {
			grid[x][y] = rand() % 2 == 0 ? '@' : ' ';
		}
	}

	for (;;) {
		print_grid(grid);
		next_generation(grid);
		printf("\e[1;1H\e[2J");
		sleep(1);
	}

	return 0;
}

int count_neighbors(int x, int y, char grid[ROWS][COLS])
{
	int d, nx, ny;
	int neighbors;

	neighbors = 0;
	for (d = 0; d < 8; d++) {
		nx = x + deltas[d][0];
		ny = y + deltas[d][1];

		if (nx < 0) {
			nx = ROWS - 1;
		} else if (nx > ROWS - 1) {
			nx = 0;
		}

		if (ny < 0) {
			ny = COLS - 1;
		} else if (ny > COLS - 1) {
			ny = 0;
		}

		if (grid[nx][ny] == '@') {
			neighbors++;
		}
	}

	return neighbors;
}

void next_generation(char grid[ROWS][COLS])
{
	int x, y, n;
	char prev_grid[ROWS][COLS];

	memcpy(prev_grid, grid, sizeof(char) * ROWS * COLS);

	for (x = 0; x < ROWS; x++) {
		for (y = 0; y < COLS; y++) {
			n = count_neighbors(x, y, prev_grid);
			if (prev_grid[x][y] == '@') {
				grid[x][y] = (n == 2 || n == 3) ? '@' : ' ';
			} else {
				grid[x][y] = (n == 3) ? '@' : ' ';
			}
		}
	}
}

void print_grid(char grid[ROWS][COLS])
{
	int x, y;

	for (x = 0; x < ROWS; x++) {
		for (y = 0; y < COLS; y++) {
			printf("%c", grid[x][y]);
		}
		printf("\n");
	}
}
